src/renderer/vt/paint.cpp

```diff
@@ -445,10 +450,36 @@ using namespace Microsoft::Console::Types;
                                      (totalWidth - numSpaces) :
                                      totalWidth;
 
+    if (cchActual == 0)
+    {
+        // If the previous row wrapped, but this line is empty, then we actually
+        // do want to move the cursor down. Otherwise, we'll possibly end up
+        // accidentally erasing the last character from the previous line, as
+        // the cursor is still waiting on that character for the next character
+        // to follow it.
+        _wrappedRow = std::nullopt;
+    }
+
+    // Move the cursor to the start of this run.
+    RETURN_IF_FAILED(_MoveCursor(coord));
+
     // Write the actual text string
     std::wstring wstr = std::wstring(unclusteredString.data(), cchActual);
     RETURN_IF_FAILED(VtEngine::_WriteTerminalUtf8(wstr));
 
+    // If we've written text to the last column of the viewport, then mark
+    // that we've wrapped this line. The next time we attempt to move the
+    // cursor, if we're trying to move it to the start of the next line,
+    // we'll remember that this line was wrapped, and not manually break the
+    // line.
+    // Don't do this is the last character we're writing is a space - The last
+    // char will always be a space, but if we see that, we shouldn't wrap.
+    if (lineWrapped &&
+        (_lastText.X + (totalWidth - numSpaces)) > _lastViewport.RightInclusive())

```

Comment:
@miniksa What's the correct new-safemath way of doing this?

Reply:
Depends if you want an error out of bounds or if you want it to clamp. In most cases, we just want to clamp, so far.

You can either turn some of these variables into `base::ClampedNumeric<T>` or `base::CheckedNumeric<T>` and then use the usual operators on them to do the math normally and extract the result at the end. Clamped will auto-extract on assignment to a normal type. Checked requires you to choose one of the three extractions (ValueOrDie, ValueOrDefault, or AssignIfValid) to determine how you want to handle it.

Alternatively, you can just do the work with the bare types using `base::ClampAdd()` or `base::CheckAdd()` in a similar fashion to the old safe math way. Your choice. Whichever looks least ugly.

Finally, there's also `base::saturating_cast<T>` and `base::checked_cast<T>` if you want to do everything some other way in a larger datatype and then cast to check at the end.

It'll be completely contextual so I'm leaving it to you to decide which way works best here as I didn't read the whole function context, just your diffs.